pyevmasm/evmasm.py

```diff
@@ -71,483 +72,283 @@ class InstructionTable(dict):
         Instruction(0x3, 'SUB', 0, 2, 1, 3, 'Subtraction operation.', None, 0)
 
     """
-
+    __slots__ = ('_previous_fork', '_cache')
     def __init__(self, *args, **kwargs):
         super(InstructionTable, self).__init__(*args, **kwargs)
-        # Don't build the reverse LUT unless we actually use it
-        self._rtable = None
-
-    def _reverse_lut(self):
-        reverse_table = {}
-        for (opcode, (name, immediate_operand_size, pops, pushes, gas, description)) in self.items():
-            mnemonic = name
-            if name == 'PUSH':
-                mnemonic = '%s%d' % (name, (opcode & 0x1f) + 1)
-            elif name in ('SWAP', 'LOG', 'DUP'):
-                mnemonic = '%s%d' % (name, (opcode & 0xf) + 1)
-            reverse_table[mnemonic] = opcode, name, immediate_operand_size, pops, pushes, gas, description
-        return reverse_table
+        self._previous_fork = kwargs.get('previous_fork', None)
+        self._cache = {}
 
-    @property
-    def reverse_table(self):
-        if not self._rtable:
-            self._rtable = self._reverse_lut()
-        return self._rtable
-
-    def __iter__(self):
-        self._k = iter(self.keys())
-        return self
-
-    def next(self):
-        # For Python 2 support
-        return self.__next__()
-
-    def __next__(self):
-        opcode = next(self._k)
-        return Instruction(opcode, *super(InstructionTable, self).__getitem__(opcode))
-
-    def __getitem__(self, item):
-        if isinstance(item, slice):
-            raise NotImplementedError
-        if isinstance(item, int):
-            try:
-                return Instruction(item, *super(InstructionTable, self).__getitem__(item))
-            except KeyError:
-                try:
-                    return Instruction(*self.reverse_table['INVALID'])
-                except KeyError:
-                    raise UnknownOpcodeError(item)
-        elif isinstance(item, str):
-            try:
-                return Instruction(*self.reverse_table[item])
-            except KeyError:
-                raise UnknownMnemonicError(item)
-
-    def __setitem__(self, key, value):
-        return
+
+    def _search_by_opcode(self, k):
+        did_not_found = object()
+        value = did_not_found
+        try:
+            value = Instruction(k, *super(InstructionTable, self).__getitem__(k))
+        except KeyError as e:
+            if self._previous_fork is not None:
+                value = self._previous_fork.get(k, did_not_found)
+
+        if value is did_not_found:
+            raise KeyError(k)
```

Comment:
Isn't it exactly what the `self._previous_fork[k]` would do if we write it like this?

It's a bit weird to use `object()` as a 'sentinel' or 'guardian'. It seems the code could be written like this:

```python
try:
    value = Instruction(k, *super(InstructionTable, self).__getitem__(k))
except KeyError as e:
    if self._previous_fork is None:
        raise KeyError(k)
    value = self._previous_fork[k]

return value
```

also maybe `*super(InstructionTable, self)[k]` instead of using `.__getitem__(k)`? Not sure if it works, tho.